### PR TITLE
Prepare to use C++23 in WebKit

### PR DESCRIPTION
**Pull request by achristensen07**
#### 272547d4b0ece68623a629726f30a54d5705f802
<pre>
Prepare to use C++23 in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=274952">https://bugs.webkit.org/show_bug.cgi?id=274952</a>
<a href="https://rdar.apple.com/129048503">rdar://129048503</a>

Reviewed by Abrar Rahman Protyasha.

This is the parts of <a href="https://github.com/WebKit/WebKit/pull/19053">https://github.com/WebKit/WebKit/pull/19053</a> that don&apos;t involve actually changing
the compiler flag, which we can&apos;t do until June 10 because of the GTK and WPE dependencies policy.

From that PR:

1. std::aligned_storage is deprecated, so I added deprecation warning ignoring macros around its use
   in order to not change any behavior with the transition.  We can figure out what to do later.
2. Something about implicitly calling constructors changed, which affected code in JSBigInt.
   I fixed this by just adding braces to more explicitly indicate that a different type is being
   returned and we are calling a constructor.  This should also not change behavior.
3. For a reason I don&apos;t completely understand, in order to fix the build I needed to change the name
   of WasmMemory::check to checkLifetime.  &quot;check&quot; is certainly not a new keyword.  This also does
   not change any behavior.
4. std::unique_ptr is now constexpr, and something in libc++&apos;s implementation requires a complete
   class definition for its move constructor.  I filed <a href="https://rdar.apple.com/116885103">rdar://116885103</a> for future investigation,
   but I also moved a few definitions around and included a few things to fix the build.
   The SerializedScriptValue constructor taking a unique_ptr&lt;WasmModuleArray&gt; was a little tricky,
   but I made it just take a WasmModuleArray&amp;&amp; instead, and call makeUnique between that and initializing
   the member variable, so I didn&apos;t change behavior or storage layout there either.
5. I needed to explicitly include &lt;iterator&gt; from EmbeddedFixedVector.h to fix the build.
   Something probably changed about what standardized headers include other standardized headers.

* Source/JavaScriptCore/assembler/ProbeStack.h:
* Source/JavaScriptCore/dfg/DFGAbstractValue.cpp:
(JSC::DFG::AbstractValue::ensureCanInitializeWithZeros):
* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::JSBigInt::exponentiateImpl):
(JSC::JSBigInt::multiplyImpl):
(JSC::JSBigInt::remainderImpl):
(JSC::JSBigInt::leftShiftImpl):
(JSC::JSBigInt::signedRightShiftImpl):
(JSC::JSBigInt::absoluteAdd):
(JSC::JSBigInt::absoluteSub):
(JSC::JSBigInt::asIntNImpl):
(JSC::JSBigInt::asUintNImpl):
* Source/JavaScriptCore/wasm/WasmMemory.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyMemory.cpp:
(JSC::JSWebAssemblyMemory::adopt):
(JSC::JSWebAssemblyMemory::growSuccessCallback):
* Source/ThirdParty/libwebrtc/Source/webrtc/modules/audio_processing/include/audio_processing.cc:
* Source/ThirdParty/libwebrtc/Source/webrtc/modules/audio_processing/include/audio_processing.h:
* Source/ThirdParty/libwebrtc/Source/webrtc/rtc_base/untyped_function.h:
* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/HashTable.h:
(WTF::KeyTraits&gt;::checkKey):
* Source/WTF/wtf/Int128.cpp:
* Source/WTF/wtf/Int128.h:
* Source/WTF/wtf/NeverDestroyed.h:
* Source/WTF/wtf/RobinHoodHashTable.h:
(WTF::SizePolicy&gt;::checkKey):
* Source/WTF/wtf/SizeLimits.cpp:
* Source/WTF/wtf/ThreadSpecific.h:
* Source/WTF/wtf/Vector.h:
* Source/WTF/wtf/threads/Signals.h:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::SerializedScriptValue::SerializedScriptValue):
(WebCore::SerializedScriptValue::create):
* Source/WebCore/bindings/js/SerializedScriptValue.h:
* Source/WebCore/dom/TreeScope.cpp:
* Source/WebCore/html/HTMLMediaElement.cpp:
* Source/WebCore/page/CaptionUserPreferences.cpp:
(WebCore::CaptionUserPreferences::createTestingModeToken):
* Source/WebCore/page/CaptionUserPreferences.h:
(WebCore::CaptionUserPreferences::createTestingModeToken): Deleted.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::GridTrackSizingAlgorithm):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebGPU/WGSL/Types.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
* Source/WebKit/Shared/Cocoa/WKObject.h:
* Source/WebKit/Shared/RemoteLayerTree/LayerProperties.h:
(WebKit::RemoteLayerBackingStoreOrProperties::RemoteLayerBackingStoreOrProperties): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStoreOrProperties::RemoteLayerBackingStoreOrProperties):
* Source/WebKit/UIProcess/WebBackForwardCache.cpp:
(WebKit::WebBackForwardCache::addEntry):
* Source/WebKit/UIProcess/WebBackForwardCacheEntry.h:
* Source/bmalloc/bmalloc/AllIsoHeaps.h:
* Source/bmalloc/bmalloc/BCompiler.h:
* Source/bmalloc/bmalloc/CryptoRandom.cpp:
* Source/bmalloc/bmalloc/DebugHeap.h:
* Source/bmalloc/bmalloc/Environment.h:
* Source/bmalloc/bmalloc/Gigacage.cpp:
* Source/bmalloc/bmalloc/HeapConstants.h:
* Source/bmalloc/bmalloc/IsoSharedHeap.h:
* Source/bmalloc/bmalloc/IsoTLSLayout.h:
* Source/bmalloc/bmalloc/PerHeapKind.h:
* Source/bmalloc/bmalloc/Scavenger.h:

Canonical link: <a href="https://commits.webkit.org/279578@main">https://commits.webkit.org/279578@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/4f7e93a0d883eb81ab3a464991b0d776523af96d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33212 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6367 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57122 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4567 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56148 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40708 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4431 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43598 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2996 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55942 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31419 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46578 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24738 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28247 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3889 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2722 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/47202 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4081 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58717 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/53355 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29014 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4140 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51010 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30206 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46721 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50346 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/memory-pressure (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31142 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65657 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7970 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29985 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12501 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/assembler/ProbeStack.h [hunk 1]
```diff
@@ -139,7 +139,9 @@ class Page {
     static_assert(s_pageSize > s_chunkSize, "bad pageSize or chunkSize");
     static_assert(s_chunkSize == (1 << s_chunkSizeShift), "bad chunkSizeShift");
 
+    ALLOW_DEPRECATED_DECLARATIONS_BEGIN
```
Comment (aprotyas) [inline Source/JavaScriptCore/assembler/ProbeStack.h:142]: Do you mind opening a bug to track migration from `std::aligned_storage`?

Comment (aprotyas) [reply Source/JavaScriptCore/assembler/ProbeStack.h:142]: It looks like there's very valid reason to drop this.

**Review by aprotyas — APPROVED** [review]
Looks good to me!

File: Source/JavaScriptCore/assembler/ProbeStack.h (hunk 1 above)
Comment (achristensen07) [reply Source/JavaScriptCore/assembler/ProbeStack.h:142]: https://bugs.webkit.org/show_bug.cgi?id=274966

**Pull request merged**